File: src/utils/ExpressionParser.php
<?php

class ExpressionParser {
    private $replaceAll = false;
    private $from = "";
    private $to = "";
    private $expression;
    private $expressionSplitted;
    private $text;

    private $delimiter = "/";

    public $result;

    function __construct(String $text, String $expression) {
        $this->text = $text;
        $this->expression = $expression;

        $this->sanitize();

        if(!$this->validadeExpression()){
            return;
        } else {
            $this->run();
        }
    }

    function sanitize() {
        $this->expressionSplitted = explode($this->delimiter, $this->expression);
        foreach($this->expressionSplitted as $key => $value) {
            if($value == "\\"){
                if($key){
                    $this->expressionSplitted[$key] = "/".$this->expressionSplitted[$key+1];
                    unset($this->expressionSplitted[$key+1]);
                }
            }
            if(empty($value)){
                unset($this->expressionSplitted[$key]);
            }
        }
        $this->expressionSplitted = array_values($this->expressionSplitted);
    }

    function validadeExpression() {
    
        if($this->expressionSplitted[0] != "s"){
            return false;
        }

        if(count($this->expressionSplitted) <= 3 && count($this->expressionSplitted) >= 4){
            return false;
        }

        return true;
    }

    function replaceAllChecker() {
        if(count($this->expressionSplitted) == 4){
            if($this->expressionSplitted[3] == "g"){
                $this->replaceAll = true;
            }
        }
    }

    function preRun() {
        $this->from = $this->expressionSplitted[1];
        $this->to = $this->expressionSplitted[2];
        $this->replaceAllChecker();
    }

    function run() {
        $this->preRun();
        
        if($this->from == "*") {
            $this->from = $this->text;
        }
        
        $this->result = $this->replace();
        
    }

    function replace() {
        if($this->replaceAll){
         
            return preg_replace("/".$this->from."/", $this->to, $this->text);
        }
        

        return preg_replace("/".$this->from."/", $this->to, $this->text ,1);
    }
}

File: lib/utils/ExpressionParser.dart
class ExpressionParser {
  static const delimiter = '/';
  String from, to;
  String expression;
  List<String> expressionSplitted;
  String text;
  bool replaceAll = false;
  bool deleteSource = false;

  String result;

  ExpressionParser({this.text, this.expression}) {
    if (expression.contains('//')) {
      deleteSource = true;
      expression = expression.replaceFirst('//', '/');
    }

    expressionSplitted = expression.split(delimiter);

    replaceAll = expressionSplitted.last == 'g';
    from = expressionSplitted.elementAt(1);
    to = expressionSplitted.elementAt(2);

    if (from == '*') {
      from = text;
    }

    result = run();
  }

  bool isValid() {
    if (expression[0] != 's') return false;
    if (!(expression.endsWith('/') || expression.endsWith('/g'))) return false;

    if (expressionSplitted.length <= 3 && expressionSplitted.length >= 4) {
      return false;
    }

    return true;
  }

  String run() {
    if (replaceAll) return text.replaceAll(from, to);

    return text.replaceFirst(from, to);
  }

  @override
  String toString() {
    return """
    delimiter: $delimiter,
    from: $from,
    to: $to,
    expression: $expression,
    expressionSplitted: ${expressionSplitted.join(', ')}
    text: $text,
    replaceAll: $replaceAll,
    deleteSource: $deleteSource

    result: $result
    """;
  }
}
